extensions/kafka-eight-simpleConsumer/src/main/java/io/druid/firehose/kafka/KafkaSimpleConsumer.java

```diff
@@ -0,0 +1,402 @@
+    return earliest ? offsets[0] : offsets[offsets.length - 1];
+  }
+
+  public Iterable<BytesMessageWithOffset> fetch(long offset, int timeoutMs) throws InterruptedException
+  {
+    FetchResponse response = null;
+    Broker previousLeader = leaderBroker;
+    while (true) {
+      ensureConsumer(previousLeader);
+
+      FetchRequest request = new FetchRequestBuilder()
+          .clientId(clientId)
+          .addFetch(topic, partitionId, offset, 100000000)
+          .maxWait(timeoutMs)
+          .minBytes(1)
+          .build();
+
+      log.debug("fetch offset %s", offset);
+
+      try {
+        response = consumer.fetch(request);
+      }
+      catch (Exception e) {
+        ensureNotInterrupted(e);
+        log.warn(e, "caughte exception in fetch {} - {}", topic, partitionId);
+        response = null;
+      }
+
+      if (response == null || response.hasError()) {
+        short errorCode = response != null ? response.errorCode(topic, partitionId) : ErrorMapping.UnknownCode();
+        log.warn("fetch %s - %s with offset %s encounters error: [%s]", topic, partitionId, offset, errorCode);
+
+        boolean needNewLeader = false;
+        if (errorCode == kafkaRequestTimeoutCode) {
+          log.info("kafka request timed out, response[%s]", response);
+        } else if (errorCode == kafkaOffsetOutOfRangeCode) {
+        	long newOffset = getOffset(earliest);
+          log.info("got [%s] offset[%s] for [%s][%s]", earliest ? "earliest" : "latest", newOffset, topic, partitionId);
+          if (newOffset < 0) {
+            needNewLeader = true;
+          } else {
+            offset = newOffset;
+            continue;
+          }
+        } else {
+          needNewLeader = true;
+        }
+
+        if (needNewLeader) {
+          stopConsumer();
+          previousLeader = leaderBroker;
+          leaderBroker = null;
+          continue;
+        }
+      } else {
+        break;
+      }
+    }
+
+    return response != null ? filterAndDecode(response.messageSet(topic, partitionId), offset) : EMPTY_MSGS;
+  }
+
+  private void stopConsumer()
+  {
+    if (consumer != null) {
+      try {
+        consumer.close();
+        log.info("stop consumer[%s][%s], leaderBroker[%s]", topic, partitionId, leaderBroker);
+      }
+      catch (Exception e) {
+        log.warn(e, "stop consumer[%s][%s] failed", topic, partitionId);
+      }
+      finally {
+        consumer = null;
+      }
+    }
+  }
+
+  // stop the consumer
+  public void stop()
+  {
+    stopConsumer();
+    log.info("KafkaSimpleConsumer[%s][%s] stopped", topic, partitionId);
+  }
+
+  private PartitionMetadata findLeader() throws InterruptedException
+  {
+    for (KafkaBroker broker : replicaBrokers) {
+      SimpleConsumer consumer = null;
+      try {
+        log.info("findLeader, try broker %s:%s", broker.host, broker.port);
+        consumer = new SimpleConsumer(broker.host, broker.port, soTimeout, bufferSize, clientId + "leaderLookup");
```

Comment:
Just to create a consumer with a different clientId than the one that consumes event for firehose. This consumer is only used for leader lookup when current Kafka broker leader goes down. I understand that this clientId is not announced to user, so I added a log in `KafkaSimpleConsumer` constructor. I wonder if that would help a little? Or should I change it in other ways?